plugin/output/s3/s3.go

```diff
@@ -96,97 +211,159 @@ func (p *Plugin) Start(config pipeline.AnyConfig, params *pipeline.OutputPluginP
 
-	dir, f := filepath.Split(p.config.FileConfig.TargetFile)
+	// dir for all bucket files.
+	p.setDirs(outPlugCount)
+	// file for each bucket.
+	p.setFileNames(outPlugCount)
 
-	p.targetDir = dir
-	p.fileExtension = filepath.Ext(f)
-	p.fileName = f[0 : len(f)-len(p.fileExtension)]
-
-	p.uploadCh = make(chan string, p.config.FileConfig.WorkersCount_*4)
-	p.compressCh = make(chan string, p.config.FileConfig.WorkersCount_)
+	p.uploadCh = make(chan fileDTO, p.config.FileConfig.WorkersCount_*4)
+	p.compressCh = make(chan fileDTO, p.config.FileConfig.WorkersCount_)
 
 	for i := 0; i < p.config.FileConfig.WorkersCount_; i++ {
 		longpanic.Go(p.uploadWork)
 		longpanic.Go(p.compressWork)
 	}
 
-	// initialize minio client object.
-	minioClient, err := minio.New(p.config.Endpoint, p.config.AccessKey, p.config.SecretKey, p.config.Secure)
-	if err != nil || minioClient == nil {
-		p.logger.Panicf("could not create minio client, error: %s", err.Error())
-	}
-	p.client = minioClient
-
-	if p.config.client != nil {
-		p.logger.Info("set mock client")
-		p.client = *p.config.client
-	}
-
-	exist, err := p.client.BucketExists(p.config.Bucket)
+	// initialize minio clients.
+	clients, err := p.minioClients()
 	if err != nil {
-		p.logger.Panicf("could not check bucket: %s, error: %s", p.config.Bucket, err.Error())
+		p.logger.Panicf("could not create minio client, error: %s", err.Error())
 	}
-	if !exist {
-		p.logger.Fatalf("bucket: %s, does not exist", p.config.Bucket)
+	p.clients = clients
+
+	err = p.startPlugins(params, outPlugCount)
+	switch err.(type) {
+	case nil:
+		break
+	case *ErrCreateOutputPluginPanic:
+		p.logger.Panic(err.Error())
+	case *ErrCreateOutputPluginNoSuchBucket:
+		p.logger.Fatal(err.Error())
 	}
-	p.logger.Info("client is ready")
-	p.logger.Infof("bucket: %s exists", p.config.Bucket)
 
-	anyPlugin, _ := file.Factory()
-	p.outPlugin = anyPlugin.(*file.Plugin)
-
-	p.outPlugin.SealUpCallback = p.addFileJob
-
-	p.outPlugin.Start(&p.config.FileConfig, params)
 	p.uploadExistingFiles()
 }
 
 func (p *Plugin) Stop() {
-	p.outPlugin.Stop()
+	p.outPlugins.Stop()
 }
 
 func (p *Plugin) Out(event *pipeline.Event) {
-	p.outPlugin.Out(event)
+	p.outPlugins.Out(event, pipeline.PlugSelector{
+		CondType:  pipeline.ByNameSelector,
+		CondValue: p.decideReceiverBucket(event),
+	})
 }
 
-// uploadExistingFiles gets files from dirs, sorts it, compresses it if it's need, and then upload to s3
-func (p *Plugin) uploadExistingFiles() {
-	// get all compressed files
-	pattern := fmt.Sprintf("%s*%s", p.targetDir, p.compressor.getExtension())
-	compressedFiles, err := filepath.Glob(pattern)
+// decideReceiverBucket decide which s3 bucket shall receive event.
+func (p *Plugin) decideReceiverBucket(event *pipeline.Event) string {
+	bucket := event.Root.Dig(p.config.BucketFieldInEvent).AsString()
+	// no BucketFieldInEvent in message, it's DefaultBucket, showtime.
+	if len(bucket) == 0 {
+		return p.config.DefaultBucket
+	}
+	// Bucket already exists, showtime.
```

Comment:
may create plugin which alredy existed. Also be race on clients and etc.
agreed to provide a separate default client and add lock here.

Reply:
After a while I went to this:
```go
func (p *Plugin) decideReceiverBucket(event *pipeline.Event) string {
	bucketName := event.Root.Dig(p.config.BucketFieldInEvent).AsString()
	// No race, because p.config.DefaultBucket immutable, RO.
	if len(bucketName) == 0 {
		return p.config.DefaultBucket
	}

	// Concurrent safety lays on p.outPlugins itself.
	if p.outPlugins.Exists(bucketName) {
		return bucketName
	}

	// Lock for creating new dynamic bucket
	p.dynamicPlugCreationMu.Lock()

	// Some one may already create bucket.
	if p.outPlugins.IsDynamic(bucketName) {
		return bucketName
	}
	// If not we will create!
	if created := p.tryRunNewPlugin(bucketName); created {
		// Succeed, return new bucketName.
		return bucketName
	}
	p.dynamicPlugCreationMu.Unlock()

	// Failed to create, fallback on DefaultBucket.
	return p.config.DefaultBucket
}
```